### PR TITLE
Persist workspace logs to S3

### LINKED ISSUE #6374: Persist workspace container logs for debugging
It's currently difficult to debug issues with workspace containers after the fact, as we don't have any way to retrieve logs from containers that no longer exist. To improve our ability to debug, we should persist workspace logs somewhere. A few ideas:

- Log to CloudWatch streams, one per workspace/version combo. Docker actually has native support for this, and it works pretty much automatically when running on EC2 (which we are): https://docs.docker.com/config/containers/logging/awslogs/
- Dump logs to S3 before we shut down a container; we can use the Docker API to get all logs.

I think CloudWatch is my preferred option here. Because it's built into Docker, we don't have to think about manually flushing logs if a container dies unexpectedly (or deliberately when we will it). It's probably more expensive than S3, but if we set an appropriate retention policy (say, 7 days), we should be able to keep costs relatively low in the long term. Note that the majority of CloudWatch pricing comes from ingestion ($0.50/GB), whereas storage is only slightly more expensive than S3 ($0.03/GB to store, vs $0.023/GB for S3).

### PR DESCRIPTION
**Pull request by nwalters512**
This PR enables the workspace host to persist logs from its containers to S3. Logs are stored according to workspace and version. This does mean that logs for rebooted containers will be in the same namespace and not be distinguishable from each other; I think that's OK?

Note that this is currently only useful for folks with infrastructure access. #6530 captures followup work to expose these logs to instructors via the web UI.

Closes #6374.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -34,6 +34,8 @@ test-prairielib:
 	@yarn jest prairielib/
 test-grader-host:
 	@yarn jest grader_host/
+test-workspace-host:
+	@yarn mocha "workspace_host/**/*.test.{js,mjs}"
```
Comment (nwalters512) [inline Makefile:38]: I'm colocating the tests with the implementation.

File: docs/docker.md [hunk 2]
```diff
@@ -93,12 +93,14 @@ The equivalent `docker run` command to perform all these actions would be:
...
-      -v /var/run/docker.sock:/var/run/docker.sock
-      prairielearn/prairielearn
+  -p 3000:3000 \
+  -v $PWD/testCourse:/course \
+  -v $HOME/pl_ag_jobs:/jobs \
+  -v $PWD:/PrairieLearn \
+  -v /var/run/docker.sock:/var/run/docker.sock \
+  -e HOST_JOBS_DIR=$HOME/pl_ag_jobs \
+  -e NODEMON=true \
+  prairielearn/prairielearn
```
Comment (nwalters512) [inline docs/docker.md:103]: The existing command used broken syntax; this was my attempt to improve it. This works for me locally now.

File: lib/config.js [hunk 3]
```diff
@@ -173,6 +173,7 @@ config.ltiRedirectUrl = null;
 config.filesRoot = '/files';
 config.trustProxy = false; // Set this to a value from (https://expressjs.com/en/4x/api.html#trust.proxy.options.table) if we're behind a proxy
 config.workspaceS3Bucket = 'workspaces';
+config.workspaceLogsS3Bucket = 'workspace-logs';
```
Comment (nwalters512) [inline lib/config.js:176]:
```suggestion
config.workspaceLogsS3Bucket = null;
```

File: workspace_host/interface.js [hunk 4]
```diff
@@ -1228,6 +1242,20 @@ function _createContainer(workspace, callback) {
...
       },
+      async () => {
+        if (config.workspaceLogsS3Bucket) {
+          const logForwarder = new ContainerS3LogForwarder(container, {
+            bucket: config.workspaceLogsS3Bucket,
+            prefix: `${workspace.id}/${workspace.version}`,
+            interval: config.workspaceLogsFlushIntervalSec * 1000,
+            tags: {
+              // TODO: tag by course and institution too?
+              WorkspaceId: workspace.id,
```
Comment (nwalters512) [inline workspace_host/interface.js:1253]: Tagging should give us some more insights into cost via Cost Explorer.

**Review by mwest1066 — APPROVED** [review]

File: workspace_host/lib/docker.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  function bufferSlice(end) {
+    const out = buffer.subarray(0, end);
+    buffer = buffer.subarray(end, buffer.length);
+    return out;
+  }
+
+  return output;
+}
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply workspace_host/lib/docker.js:38]: Logs were actually *never* printed to the console to begin with. That might be a nice improvement, but I'd leave that for another PR.

File: workspace_host/interface.js [hunk 6]
```diff
@@ -1228,6 +1242,20 @@ function _createContainer(workspace, callback) {
...
         );
       },
+      async () => {
+        if (config.workspaceLogsS3Bucket) {
+          const logForwarder = new ContainerS3LogForwarder(container, {
+            bucket: config.workspaceLogsS3Bucket,
+            prefix: `${workspace.id}/${workspace.version}`,
+            interval: config.workspaceLogsFlushIntervalSec * 1000,
+            tags: {
+              // TODO: tag by course and institution too?
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply workspace_host/interface.js]: Done in 45fa275 - can you sanity-check?

File: workspace_host/lib/docker.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mwest1066) [reply workspace_host/lib/docker.js:38]: Oh! In that case this is a complete win for observability. Great!

[the remark this answers is not in the document]
Comment (mwest1066) [reply workspace_host/lib/docker.js:38]: Oh! In that case this is a complete win for observability. Great!

File: workspace_host/interface.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mwest1066) [reply workspace_host/interface.js]: lgtm!

**Pull request merged**